pkg/drivers/vfkit/vfkit.go

```diff
@@ -0,0 +1,513 @@
+	var isoPath = filepath.Join(machineDir, isoFilename)
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-blk,path=%s", isoPath))
+
+	var mac = d.MACAddress
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-net,nat,mac=%s", mac))
+
+	startCmd = append(startCmd,
+		"--device", "virtio-rng")
+
+	startCmd = append(startCmd,
+		"--kernel", d.ResolveStorePath("bzimage"))
+	startCmd = append(startCmd,
+		"--kernel-cmdline", d.Cmdline)
+	startCmd = append(startCmd,
+		"--initrd", d.ResolveStorePath("initrd"))
+
+	for i := 0; i < d.ExtraDisks; i++ {
+		startCmd = append(startCmd,
+			"--device", fmt.Sprintf("virtio-blk,path=%s", pkgdrivers.ExtraDiskPath(d.BaseDriver, i)))
+	}
+
+	startCmd = append(startCmd,
+		"--device", fmt.Sprintf("virtio-blk,path=%s", d.diskPath()))
+
+	log.Debugf("executing: vfkit %s", strings.Join(startCmd, " "))
+	os.Remove(d.sockfilePath())
+	cmd := exec.Command("vfkit", startCmd...)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	pid := cmd.Process.Pid
+	if err := os.WriteFile(d.pidfilePath(), []byte(fmt.Sprintf("%v", pid)), 0600); err != nil {
+		return err
+	}
+
+	// Need to strip 0's
+	mac = pkgdrivers.TrimMacAddress(mac)
+	if err := d.setupIP(mac); err != nil {
+		return err
+	}
+
+	log.Infof("Waiting for VM to start (ssh -p %d docker@%s)...", d.SSHPort, d.IPAddress)
+
+	return WaitForTCPWithDelay(fmt.Sprintf("%s:%d", d.IPAddress, d.SSHPort), time.Second)
+}
+
+func (d *Driver) setupIP(mac string) error {
+	var err error
+
+	getIP := func() error {
+		d.IPAddress, err = pkgdrivers.GetIPAddressByMACAddress(mac)
+		if err != nil {
+			return errors.Wrap(err, "failed to get IP address")
+		}
+		return nil
+	}
+
+	// Implement a retry loop because IP address isn't added to dhcp leases file immediately
+	for i := 0; i < 60; i++ {
+		log.Debugf("Attempt %d", i)
+		err = getIP()
+		if err == nil {
+			break
+		}
+		time.Sleep(2 * time.Second)
+	}
+
+	if err != nil {
+		return fmt.Errorf("IP address never found in dhcp leases file %v", err)
+	}
+	log.Debugf("IP: %s", d.IPAddress)
+
+	return nil
+}
+
+func (d *Driver) Stop() error {
+	if err := d.SetVFKitState("HardStop"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Driver) Remove() error {
+	s, err := d.GetState()
+	if err != nil {
+		return errors.Wrap(err, "get state")
+	}
+	if s == state.Running {
+		if err := d.Kill(); err != nil {
+			return errors.Wrap(err, "kill")
+		}
+	}
+	if s != state.Stopped {
+		if err := d.SetVFKitState("Stop"); err != nil {
+			return errors.Wrap(err, "quit")
+		}
+	}
+	return nil
+}
+
+func (d *Driver) Restart() error {
+	s, err := d.GetState()
+	if err != nil {
+		return err
+	}
+
+	if s == state.Running {
+		if err := d.Stop(); err != nil {
+			return err
+		}
+	}
+	return d.Start()
+}
+
+func (d *Driver) extractKernel(isoPath string) error {
+	for _, f := range []struct {
+		pathInIso string
+		destPath  string
+	}{
+		{"/boot/bzimage", "bzimage"},
```

Comment:
On arm64, vfkit requires an uncompressed kernel, I don't know what format `bzimage` is in.
Also be aware that there were some issues with direct kernel boot both on amd64 and arm64 on macOS 12 https://github.com/crc-org/vfkit/issues/11

If you can require macOS 13+, I'd recommend using uefi boot if possible, no need to extract kernel/initrd/...

Reply:
For some reason it worked fine with boot2docker.iso, but failed to boot with minikube.iso

It might be the extra copy of the kernel, so might require some ISO cleanup before it works?

Dropping support for macOS 12 is not so much of an issue, I think? Had to upgrade anyway

Reply:
The kernel was _supposed_ to be compressed with lz4 (at least so I thought), but seems uncompressed

Could be some old bug with the aarch64 iso build perhaps, when it is copying the wrong file or something

Reply:
> Also be aware that there were some issues with direct kernel boot both on amd64 and arm64 on macOS 12 https://github.com/crc-org/vfkit/issues/11

Minikube still uses kernel 5.10, so that should be OK. Perhaps going to upgrade to 5.15, but still (< 5.18)

Reply:
I think for new code we can assume current macOS version as requirement. The only reason to run older macOS is old hardware that cannot be upgraded (6+ years old), and such hardware is typically not usable for running minikube with virtual machines.